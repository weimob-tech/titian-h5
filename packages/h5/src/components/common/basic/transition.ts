import { isPlainObject } from '../utils';
import { raf } from '../utils/raf';

export interface Timeout {
  exit?: number;
  appear?: number;
}

enum TransitionStatus {
  enter,
  exit,
}

const DEFAULT_TIMEOUT = 300;

export type TransitionName =
  | 'fade'
  | 'fade-down'
  | 'fade-up'
  | 'fade-left'
  | 'fade-right'
  | 'slide-up'
  | 'slide-down'
  | 'slide-left'
  | 'slide-right'
  | 'zoom';

const getTransitions = (name: string) => {
  const transitionsMapping = new Map();

  transitionsMapping.set('enter', `titian-${name}-enter titian-${name}-enter-active enterClass enterActiveClass`);
  transitionsMapping.set(
    'enter-done',
    `titian-${name}-enter-done titian-${name}-enter-active enterDoneClass enterActiveClass`,
  );
  transitionsMapping.set('exit', `titian-${name}-exit titian-${name}-exit-active exitClass exitActiveClass`);
  transitionsMapping.set(
    'exit-done',
    `titian-${name}-exit-done titian-${name}-exit-active exitDoneClass exitActiveClass`,
  );

  return transitionsMapping;
};
export default class Transition {
  [x: string]: any;

  get show() {
    return this.$show;
  }

  set show(value: boolean) {
    this.$show = value;
    this.updateShow(value);
  }

  get timeout() {
    return this.$timeout || DEFAULT_TIMEOUT;
  }

  set timeout(value: number | Timeout) {
    this.$timeout = value;
  }

  get name() {
    return this.$name || 'fade';
  }

  set name(value: TransitionName) {
    this.$name = value;
  }

  set enterName(value: TransitionName) {
    this.$enterName = value;
  }

  get enterName() {
    return this.$enterName || this.name;
  }

  set exitName(value: TransitionName) {
    this.$exitName = value;
  }

  get exitName() {
    return this.$exitName || this.name;
  }

  set destroyOnExit(value: boolean) {
    this.$destroyOnExit = value;
  }

  get destroyOnExit() {
    return this.$destroyOnExit || this.name;
  }

  set timingFunction(value: string) {
    this.$timingFunction = value;
  }

  get timingFunction() {
    return this.$timingFunction || 'linear';
  }

  set classes(value: string) {
    this.$classes = value;
  }

  get classes() {
    return this.$classes || '';
  }

  set display(value: boolean) {
    this.$display = value;
  }

  get display() {
    return this.$display || false;
  }

  set transitionEnd(value: boolean) {
    this.$transitionEnd = value;
  }

  get transitionEnd() {
    return this.$transitionEnd;
  }

  set status(value: TransitionStatus) {
    this.$status = value;
  }

  get status() {
    return this.$status;
  }

  updateShow(newValue: boolean, oldValue?: boolean) {
    if (newValue !== oldValue) {
      if (newValue) {
        this.enterHandler();
      } else {
        this.exitHandler();
      }
    }
  }

  private enterHandler() {
    const { name, enterName, timeout } = this;

    const duration = isPlainObject(timeout) ? timeout.appear : timeout;
    const transitionsMapping = getTransitions(enterName || name);
    this.status = TransitionStatus.enter;

    if (this.enterEvent && this.enterEvent.emit) {
      this.enterEvent.emit();
    }

    raf(() => {
      if (this.status !== TransitionStatus.enter) {
        return;
      }

      if (this.enteringEvent && this.enteringEvent.emit) {
        this.enteringEvent.emit();
      }

      this.duration = duration;
      this.initialized = true;
      this.display = true;
      this.classes = transitionsMapping.get('enter');
      raf(() => {
        if (this.status !== TransitionStatus.enter) {
          return;
        }
        this.transitionEnd = false;
        this.classes = transitionsMapping.get('enter-done');
        setTimeout(() => this.onTransitionEnd(), duration as number);
      });
    });
  }

  private exitHandler() {
    const { display, exitName, name, timeout } = this;
    if (!display) return;

    const duration = isPlainObject(timeout) ? timeout.exit || DEFAULT_TIMEOUT : timeout;
    const transitionsMapping = getTransitions(exitName || name);

    this.status = TransitionStatus.exit;

    if (this.exitEvent && this.exitEvent.emit) {
      this.exitEvent.emit();
    }

    raf(() => {
      if (this.status !== TransitionStatus.exit) {
        return;
      }

      if (this.exitingEvent && this.exitingEvent.emit) {
        this.exitingEvent.emit();
      }
      this.classes = transitionsMapping.get('exit');
      this.duration = duration;

      raf(() => {
        if (this.status !== TransitionStatus.exit) {
          return;
        }
        this.transitionEnd = false;
        this.classes = transitionsMapping.get('exit-done');
        setTimeout(() => this.onTransitionEnd(), duration as number);
      });
    });
  }

  private onTransitionEnd() {
    const { display, destroyOnExit, show } = this;

    if (this.transitionEnd) return;

    this.transitionEnd = true;
    if (this.status === TransitionStatus.enter) {
      if (this.enteredEvent && this.enteredEvent.emit) {
        this.enteredEvent.emit();
      }
    } else if (this.exitedEvent && this.exitedEvent.emit) {
      this.exitedEvent.emit();
    }

    if (!show && display) {
      this.classes = '';
      this.display = false;
      if (destroyOnExit) {
        this.initialized = false;
      }
    }
  }
}
