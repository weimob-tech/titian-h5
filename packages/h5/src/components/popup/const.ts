enum EPosition {
  TOP = 'top',
  BOTTOM = 'bottom',
  LEFT = 'left',
  RIGHT = 'right',
  CENTER = 'center',
  // ABSOLUTE = 'absolute'
}

const ETransitionClass: {
  [key in EPosition]: string;
} = {
  [EPosition.TOP]: 'slide-down',
  [EPosition.BOTTOM]: 'slide-up',
  [EPosition.LEFT]: 'slide-left',
  [EPosition.RIGHT]: 'slide-right',
  [EPosition.CENTER]: 'fade',
  // [EPosition.ABSOLUTE]: 'slide-up'
};

export { EPosition, ETransitionClass };
