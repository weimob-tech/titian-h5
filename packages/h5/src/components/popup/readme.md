# ti-popup



<!-- Auto Generated Below -->


## Properties

| Property          | Attribute           | Description      | Type                                                                                                                                         | Default                   |
| ----------------- | ------------------- | ---------------- | -------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------- |
| `closeOnMask`     | `close-on-mask`     |                  | `boolean`                                                                                                                                    | `true`                    |
| `contentZIndex`   | `content-z-index`   |                  | `number`                                                                                                                                     | `zIndexTool.getZIndex()`  |
| `destroyOnClose`  | `destroy-on-close`  |                  | `boolean`                                                                                                                                    | `false`                   |
| `extClass`        | `ext-class`         | 额外的类名，添加到根节点的元素上 | `string`                                                                                                                                     | `undefined`               |
| `extContentClass` | `ext-content-class` |                  | `string`                                                                                                                                     | `undefined`               |
| `extContentStyle` | `ext-content-style` |                  | `string \| { [key: string]: string; }`                                                                                                       | `undefined`               |
| `extMaskClass`    | `ext-mask-class`    |                  | `string`                                                                                                                                     | `undefined`               |
| `extMaskStyle`    | --                  |                  | `{}`                                                                                                                                         | `{}`                      |
| `extStyle`        | `ext-style`         | 额外的样式            | `string \| { [key: string]: string; }`                                                                                                       | `undefined`               |
| `hasMask`         | `has-mask`          |                  | `boolean`                                                                                                                                    | `true`                    |
| `maskZIndex`      | `mask-z-index`      |                  | `number`                                                                                                                                     | `zIndexTool.getZIndex()`  |
| `position`        | `position`          |                  | `string`                                                                                                                                     | `'center'`                |
| `preventScroll`   | `prevent-scroll`    |                  | `boolean`                                                                                                                                    | `false`                   |
| `radius`          | `radius`            |                  | `number \| string`                                                                                                                           | `''`                      |
| `safeArea`        | `safe-area`         |                  | `boolean`                                                                                                                                    | `true`                    |
| `timeout`         | `timeout`           |                  | `Timout \| number`                                                                                                                           | `300`                     |
| `timingFunction`  | `timing-function`   |                  | `string`                                                                                                                                     | `DEFAULT_TIMING_FUNCTION` |
| `transition`      | `transition`        |                  | `"fade" \| "fade-down" \| "fade-left" \| "fade-right" \| "fade-up" \| "slide-down" \| "slide-left" \| "slide-right" \| "slide-up" \| "zoom"` | `undefined`               |
| `visible`         | `visible`           |                  | `boolean`                                                                                                                                    | `false`                   |


## Events

| Event     | Description | Type                |
| --------- | ----------- | ------------------- |
| `close`   |             | `CustomEvent<null>` |
| `enter`   |             | `CustomEvent<null>` |
| `entered` |             | `CustomEvent<null>` |
| `exit`    |             | `CustomEvent<null>` |
| `exited`  |             | `CustomEvent<null>` |
| `show`    |             | `CustomEvent<null>` |


## Dependencies

### Used by

 - [ti-action-sheet](../action-sheet)
 - [ti-calendar](../calendar)
 - [ti-dialog](../dialog)
 - [ti-dropdown-item](../dropdown-item)
 - [ti-share-sheet](../share-sheet)
 - [ti-toast](../toast)

### Depends on

- [ti-transition](../transition)
- [ti-safe-area](../safe-area)

### Graph
```mermaid
graph TD;
  ti-popup --> ti-transition
  ti-popup --> ti-safe-area
  ti-action-sheet --> ti-popup
  ti-calendar --> ti-popup
  ti-dialog --> ti-popup
  ti-dropdown-item --> ti-popup
  ti-share-sheet --> ti-popup
  ti-toast --> ti-popup
  style ti-popup fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
