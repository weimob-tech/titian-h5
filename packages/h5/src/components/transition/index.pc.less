@import '../common/static/utils';
@import '../common/static/mixins/index.less';
@import '../common/static/var.less';

@transition: ~'@{project}-transition';

@enter: ~'enter';
@enterTo: ~'enter-done';
@exit: ~'exit';
@exitTo: ~'exit-done';

@fade: ~'fade';
@fadeInDown: ~'fade-down';
@fadeInUp: ~'fade-up';
@fadeInLeft: ~'fade-left';
@fadeInRight: ~'fade-right';
@slideInUp: ~'slide-up';
@slideInDown: ~'slide-down';
@slideInRight: ~'slide-right';
@slideInLeft: ~'slide-left';
@zoom: ~'zoom';

:host(.@{project}-@{fade}-@{enter}),
:host(.@{project}-@{fade}-@{exitTo}),
:host(.@{project}-@{fadeInDown}-@{enter}),
:host(.@{project}-@{fadeInDown}-@{exitTo}),
:host(.@{project}-@{fadeInUp}-@{enter}),
:host(.@{project}-@{fadeInUp}-@{exitTo}),
:host(.@{project}-@{fadeInLeft}-@{enter}),
:host(.@{project}-@{fadeInLeft}-@{exitTo}),
:host(.@{project}-@{fadeInRight}-@{enter}),
:host(.@{project}-@{fadeInRight}-@{exitTo}) {
  opacity: 0;
}

:host(.@{project}-@{fade}-@{enter}-active),
:host(.@{project}-@{fade}-@{exit}-active) {
  transition-property: opacity;
}

:host(.@{project}-@{fadeInDown}-@{enter}-active),
:host(.@{project}-@{fadeInDown}-@{exit}-active),
:host(.@{project}-@{fadeInUp}-@{enter}-active),
:host(.@{project}-@{fadeInUp}-@{exit}-active),
:host(.@{project}-@{fadeInLeft}-@{enter}-active),
:host(.@{project}-@{fadeInLeft}-@{exit}-active),
:host(.@{project}-@{fadeInRight}-@{enter}-active),
:host(.@{project}-@{fadeInRight}-@{exit}-active),
:host(.@{project}-@{zoom}-@{enter}-active),
:host(.@{project}-@{zoom}-@{exit}-active) {
  transition-property: transform, opacity;
}

:host(.@{project}-@{slideInUp}-@{enter}-active),
:host(.@{project}-@{slideInUp}-@{exit}-active),
:host(.@{project}-@{slideInDown}-@{enter}-active),
:host(.@{project}-@{slideInDown}-@{exit}-active),
:host(.@{project}-@{slideInRight}-@{enter}-active),
:host(.@{project}-@{slideInRight}-@{exit}-active),
:host(.@{project}-@{slideInLeft}-@{enter}-active),
:host(.@{project}-@{slideInLeft}-@{exit}-active) {
  transition-property: transform, visibility;
}
:host(.@{project}-@{slideInUp}-@{enter}),
:host(.@{project}-@{slideInUp}-@{exitTo}) {
  visibility: visible;

  transform: translate3d(0, 100%, 0);
}

:host(.@{project}-@{slideInDown}-@{enter}),
:host(.@{project}-@{slideInDown}-@{exitTo}) {
  visibility: visible;

  transform: translate3d(0, -100%, 0);
}

:host(.@{project}-@{slideInRight}-@{enter}),
:host(.@{project}-@{slideInRight}-@{exitTo}) {
  visibility: visible;

  transform: translate3d(-100%, 0, 0);
}

:host(.@{project}-@{slideInLeft}-@{enter}),
:host(.@{project}-@{slideInLeft}-@{exitTo}) {
  visibility: visible;

  transform: translate3d(100%, 0, 0);
}

:host(.@{project}-@{fadeInLeft}-@{enter}),
:host(.@{project}-@{fadeInLeft}-@{exitTo}) {
  transform: translate3d(100%, 0, 0);
}

:host(.@{project}-@{fadeInRight}-@{enter}),
:host(.@{project}-@{fadeInRight}-@{exitTo}) {
  transform: translate3d(-100%, 0, 0);
}

:host(.@{project}-@{fadeInUp}-@{enter}),
:host(.@{project}-@{fadeInUp}-@{exitTo}) {
  transform: translate3d(0, 100%, 0);
}

:host(.@{project}-@{fadeInDown}-@{enter}),
:host(.@{project}-@{fadeInDown}-@{exitTo}) {
  transform: translate3d(0, -100%, 0);
}

:host(.@{project}-@{zoom}-@{enter}),
:host(.@{project}-@{zoom}-@{exitTo}) {
  transform: scale3d(0.1, 0.1, 0.1);

  opacity: 0;
}

:host(.@{transition}) {
  display: block;
}
