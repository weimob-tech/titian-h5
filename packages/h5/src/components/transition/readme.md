# ti-transition



<!-- Auto Generated Below -->


## Properties

| Property           | Attribute            | Description      | Type                                                                                                                                         | Default           |
| ------------------ | -------------------- | ---------------- | -------------------------------------------------------------------------------------------------------------------------------------------- | ----------------- |
| `customStyle`      | --                   |                  | `{}`                                                                                                                                         | `{}`              |
| `destroyOnExit`    | `destroy-on-exit`    |                  | `boolean`                                                                                                                                    | `false`           |
| `enterActiveClass` | `enter-active-class` |                  | `string`                                                                                                                                     | `''`              |
| `enterClass`       | `enter-class`        |                  | `string`                                                                                                                                     | `''`              |
| `enterDoneClass`   | `enter-done-class`   |                  | `string`                                                                                                                                     | `''`              |
| `enterName`        | `enter-name`         |                  | `string`                                                                                                                                     | `undefined`       |
| `exitActiveClass`  | `exit-active-class`  |                  | `string`                                                                                                                                     | `''`              |
| `exitClass`        | `exit-class`         |                  | `string`                                                                                                                                     | `''`              |
| `exitName`         | `exit-name`          |                  | `string`                                                                                                                                     | `undefined`       |
| `extClass`         | `ext-class`          | 额外的类名，添加到根节点的元素上 | `string`                                                                                                                                     | `''`              |
| `name`             | `name`               |                  | `"fade" \| "fade-down" \| "fade-left" \| "fade-right" \| "fade-up" \| "slide-down" \| "slide-left" \| "slide-right" \| "slide-up" \| "zoom"` | `'fade'`          |
| `show`             | `show`               |                  | `boolean`                                                                                                                                    | `false`           |
| `timeout`          | `timeout`            |                  | `Timout \| number`                                                                                                                           | `DEFAULT_TIMEOUT` |
| `timingFunction`   | `timing-function`    |                  | `string`                                                                                                                                     | `'linear'`        |


## Events

| Event      | Description | Type                |
| ---------- | ----------- | ------------------- |
| `enter`    |             | `CustomEvent<void>` |
| `entered`  |             | `CustomEvent<void>` |
| `entering` |             | `CustomEvent<void>` |
| `exit`     |             | `CustomEvent<void>` |
| `exited`   |             | `CustomEvent<void>` |
| `exiting`  |             | `CustomEvent<void>` |


## Dependencies

### Used by

 - [ti-popup](../popup)
 - [ti-preview](../preview)

### Graph
```mermaid
graph TD;
  ti-popup --> ti-transition
  ti-preview --> ti-transition
  style ti-transition fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
