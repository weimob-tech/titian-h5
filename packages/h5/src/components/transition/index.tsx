import { Component, h, Prop, State, EventEmitter, Event, Watch, Element, Host } from '@stencil/core';
import { BasicComponentAbstract, use } from '../common/basic/BasicComponent';
import Transition, { TransitionName } from '../common/basic/transition';
import { addShadowRootStyle } from '../common/utils';
import * as namespace from '../common/utils/namespace';

export interface Timout {
  exit?: number;
  appear?: number;
}

const DEFAULT_TIMEOUT = 300;

@Component({
  tag: 'ti-transition',
  styleUrls: {
    pc: 'index.pc.less',
    h5: 'index.h5.less',
  },
  shadow: true,
})
export class TiTransition implements BasicComponentAbstract {
  [x: string]: any;

  @Element() host: HTMLElement;

  @Prop() extCss = '';

  @Prop() show = false;

  @Prop() extClass = '';

  @Prop() enterClass = '';

  @Prop() enterActiveClass = '';

  @Prop() enterDoneClass = '';

  @Prop() exitClass = '';

  @Prop() exitActiveClass = '';

  @Prop() exitDoneClass = '';

  @Prop() extStyle = {};

  @Prop() timeout: Timout | number = DEFAULT_TIMEOUT;

  @Prop() name: TransitionName = 'fade';

  @Prop() enterName: string;

  @Prop() exitName: string;

  @Prop() destroyOnExit = false;

  @Prop() timingFunction = 'linear';

  @use(Transition) this;

  @State() classes = '';

  @State() display = false;

  @Event({ eventName: 'enter', bubbles: false, composed: false }) enterEvent: EventEmitter<void>;

  @Event({ eventName: 'entering', composed: false, bubbles: false }) enteringEvent: EventEmitter<void>;

  @Event({ eventName: 'entered', bubbles: false, composed: false }) enteredEvent: EventEmitter<void>;

  @Event({ eventName: 'exit', bubbles: false, composed: false }) exitEvent: EventEmitter<void>;

  @Event({ eventName: 'exiting', composed: false, bubbles: false }) exitingEvent: EventEmitter<void>;

  @Event({ eventName: 'exited', bubbles: false, composed: false }) exitedEvent: EventEmitter<void>;

  componentWillLoad(): void | Promise<void> {
    if (this.show) {
      this.toggleShow(true);
    }

    addShadowRootStyle.call(this);
  }

  @Watch('show')
  private toggleShow(newValue: boolean, oldValue?: boolean) {
    this.updateShow(newValue, oldValue);
  }

  private getClassName() {
    const { classes } = this;
    const classList = [namespace.join('transition'), classes];

    return classList.join(' ');
  }

  private completedStyle() {
    const { duration, timingFunction, display, extStyle } = this;

    let style: Record<string, any> = { ...extStyle };

    if (display) {
      style = {
        '-webkit-transition-duration': `${duration}ms`,
        'transition-duration': `${duration}ms`,
        '-webkit-transition-timing-function': timingFunction,
        'transition-timing-function': timingFunction,
        ...style,
      };
    } else {
      style.display = 'none';
    }
    return style;
  }

  render() {
    const classNames = this.getClassName();
    const parts = classNames
      .split(' ')
      .filter(i => !i.startsWith(namespace.PREFIX))
      .map(i => this[i])
      .concat(this.extClass)
      .join(' ')
      .trim();
    return (
      <Host part={parts} class={`${this.getClassName()} ${parts}`} style={this.completedStyle()}>
        {this.initialized ? <slot /> : null}
      </Host>
    );
  }
}
