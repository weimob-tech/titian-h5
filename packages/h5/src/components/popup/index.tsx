/* eslint-disable class-methods-use-this */
import { Component, EventEmitter, h, Prop, Event, Watch, Element, Host } from '@stencil/core';
import { JSXBase } from '@stencil/core/internal';

import { BasicComponentAbstract } from '../common/basic/BasicComponent';
import { Timout, TransitionName } from '../common/basic/transition';
import { stringToAttrStyle, addShadowRootStyle } from '../common/utils';

import { join, handle } from '../common/utils/namespace';
import addUnit from '../common/utils/suffix';
import zIndexTool from '../common/utils/zIndexTool';

import { EPosition, ETransitionClass } from './const';

const DEFAULT_TIMING_FUNCTION = 'cubic-bezier(0.5, 1, 0.89, 1)';

@Component({
  tag: 'ti-popup',
  styleUrls: {
    pc: 'index.pc.less',
    h5: 'index.h5.less',
  },
  shadow: true,
})
export class TiPopup implements BasicComponentAbstract {
  @Element() host: HTMLElement;

  @Prop() extCss = '';

  @Prop() extClass?: string;

  @Prop() extMaskClass?: string;

  @Prop() extContentClass?: string;

  @Prop() extStyle?: string | JSXBase.HTMLAttributes<Record<string, unknown>>['style'];

  @Prop() maskZIndex?: number = zIndexTool.getZIndex();

  @Prop() contentZIndex?: number = zIndexTool.getZIndex();

  @Prop() hasMask?: boolean = true;

  @Prop() safeArea?: boolean = true;

  @Prop() visible?: boolean = false;

  @Prop() preventScroll?: boolean = true;

  @Prop() disableGlobalTouchMove?: boolean = false;

  @Prop() closeOnMask?: boolean = true;

  @Prop() position?: `${EPosition}` = 'center';

  @Prop() radius?: string | number = '';

  @Prop() timeout?: number | Timout;

  @Prop() timingFunction? = '';

  @Prop() extMaskStyle = {};

  @Prop() extContentStyle?: string | JSXBase.HTMLAttributes<Record<string, unknown>>['style'];

  @Prop() transition?: TransitionName;

  @Prop() destroyOnClose?: boolean = false;

  @Event({ eventName: 'show', bubbles: false, composed: false }) showEvent: EventEmitter<void>;

  @Event({ eventName: 'close', bubbles: false, composed: false }) closeEvent: EventEmitter<void>;

  @Event({ eventName: 'enter', bubbles: false, composed: false }) enterEvent: EventEmitter<void>;

  @Event({ eventName: 'entered', bubbles: false, composed: false }) enteredEvent: EventEmitter<void>;

  @Event({ eventName: 'exit', bubbles: false, composed: false }) exitEvent: EventEmitter<void>;

  @Event({ eventName: 'exited', bubbles: false, composed: false }) exitedEvent: EventEmitter<void>;

  @Watch('visible')
  visibleChanged(newVal, oldVal) {
    if (oldVal === newVal) return;
    if (newVal) {
      this.onShow();
      if (this.preventScroll) {
        document.body.style.overflow = 'hidden';
      }
      if (this.disableGlobalTouchMove) {
        this.addEventListener();
      }
    } else {
      this.onClose();
      if (this.preventScroll) {
        document.body.style.overflow = 'auto';
      }
      if (this.disableGlobalTouchMove) {
        this.removeEventListener();
      }
    }
  }

  @Watch('disableGlobalTouchMove')
  disableGlobalTouchMoveChanged(newVal, oldVal) {
    if (oldVal === newVal) return;
    if (this.visible) {
      if (newVal) {
        this.addEventListener();
      } else {
        this.removeEventListener();
      }
    } else {
      this.removeEventListener();
    }
  }

  componentWillLoad() {
    addShadowRootStyle.call(this);
  }

  componentDidLoad() {
    if (this.visible) {
      if (this.preventScroll) {
        document.body.style.overflow = 'hidden';
      }
      if (this.disableGlobalTouchMove) {
        this.addEventListener();
      }
    }
  }

  disconnectedCallback() {
    this.visible = false;
    if (this.preventScroll) {
      document.body.style.overflow = 'auto';
    }
    if (this.disableGlobalTouchMove) {
      this.removeEventListener();
    }
  }

  private onShow() {
    // 弹出层打开时触发
    if (this.visible) {
      this.showEvent.emit();
    }
  }

  private onClose() {
    // 弹出层关闭时触发
    if (this.visible) {
      this.closeEvent.emit();
      this.visible = false;
    }
  }

  onTapMask = () => {
    const { closeOnMask } = this;
    if (closeOnMask) {
      this.onClose();
    }
  };

  onEnter = () => {
    this.enterEvent.emit();
  };

  onEntered = () => {
    this.enteredEvent.emit();
  };

  onExit = () => {
    this.exitEvent.emit();
  };

  onExited = () => {
    this.exitedEvent.emit();
  };

  onTapContent = (event: MouseEvent) => {
    const target = event.target as HTMLElement;
    const root = target.getRootNode();
    if (root === this.host.shadowRoot) {
      event.stopPropagation();
    }
  };

  onTouchMove(event: TouchEvent) {
    event.stopPropagation();
    if (event.cancelable) {
      event.preventDefault();
    }
  }

  onTouchMoveByWindow(event: TouchEvent) {
    if (event.cancelable) {
      event.preventDefault();
    }
  }

  addEventListener() {
    window.addEventListener('touchmove', this.onTouchMoveByWindow, { passive: false });
  }

  removeEventListener() {
    window.removeEventListener('touchmove', this.onTouchMoveByWindow);
  }

  render() {
    const { position, contentZIndex, radius, extContentStyle } = this;
    let { timeout } = this;
    let transitionClass: string = this.transition;
    if (!transitionClass) {
      transitionClass = ETransitionClass[position];
    }
    if (position === EPosition.CENTER && transitionClass !== 'fade') {
      transitionClass = `${transitionClass}-center`;
    }
    const className = join('popup', [position === 'center' && transitionClass !== 'fade' ? 'center' : position]);
    const contentStyle = {};
    if (timeout) {
      if (typeof timeout === 'number') {
        contentStyle['--private-start-duration'] = `${timeout}ms`;
        contentStyle['--private-end-duration'] = `${timeout}ms`;
      } else if (typeof timeout === 'object' && timeout.appear && timeout.exit) {
        contentStyle['--private-start-duration'] = `${timeout.appear}ms`;
        contentStyle['--private-end-duration'] = `${timeout.exit}ms`;
      }
    } else {
      timeout = {
        appear: 300,
        exit: 200,
      };
    }
    const style = {
      'z-index': contentZIndex,
      '--popup-radius': addUnit(radius),
      ...contentStyle,
      ...stringToAttrStyle(extContentStyle),
    };
    return (
      <Host>
        <ti-transition
          extStyle={{ 'z-index': this.maskZIndex || '', ...this.extMaskStyle }}
          extClass={`${join('popup-mask')} ${this.extMaskClass || ''}`}
          show={this.hasMask && this.visible}
          enterDoneClass={join('popup-enter-done')}
          exitClass={join('popup-enter-done')}
          destroyOnExit={this.destroyOnClose}
          onClick={this.onTapMask}
          timeout={timeout}
          timing-function={DEFAULT_TIMING_FUNCTION}
          name="fade"
          onTouchMove={this.onTouchMove}
        />
        <ti-transition
          show={this.visible}
          name={'' as TransitionName}
          destroyOnExit={this.destroyOnClose}
          extStyle={style}
          extClass={`${this.extContentClass || ''}  ${className}`}
          timeout={timeout}
          timingFunction={this.timingFunction}
          onClick={this.onTapContent}
          enterClass={handle('popup', [`${transitionClass}-start`])}
          enterDoneClass={handle('popup', [`${transitionClass}-middle`])}
          enterActiveClass={handle('popup', [`${transitionClass}-start-active`])}
          exitClass={handle('popup', [`${transitionClass}-middle`])}
          exitDoneClass={handle('popup', [`${transitionClass}-end`])}
          exitActiveClass={handle('popup', [`${transitionClass}-end-active`])}
          onEnter={this.onEnter}
          onEntered={this.onEntered}
          onExit={this.onExit}
          onExited={this.onExited}
        >
          <div
            part={`${this.extClass}`}
            class={` ${join('popup-box', [this.position])} ${this.extClass}`}
            style={stringToAttrStyle(this.extStyle)}
          >
            {this.position === EPosition.TOP && this.safeArea && <ti-safe-area class="noscroll" position="top" />}
            <slot />
            {this.position === EPosition.BOTTOM && this.safeArea && <ti-safe-area class="noscroll" position="bottom" />}
          </div>
        </ti-transition>
      </Host>
    );
  }
}
