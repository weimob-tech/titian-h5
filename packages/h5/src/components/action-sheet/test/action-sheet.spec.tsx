import { newSpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { TiActionSheet } from '../index';
import { TiPopup } from '../../popup';
import { TiTransition } from '../../transition';
const actions = [
  { name: '选项1', description: '描述1' },
  { name: '选项1', description: '描述1', icon: 'home' },
  { name: '选项1', description: '描述1', disabled: true },
  { name: '选项1', description: '描述1', loading: true },
];
describe('渲染 ti-action-sheet', () => {
  it('基本渲染', async () => {
    const page = await newSpecPage({
      components: [TiActionSheet],
      template: () => <ti-action-sheet title="标题" actions={actions} />,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('字段别名', async () => {
    const options = [{ label: '选项1' }];
    const page = await newSpecPage({
      components: [TiActionSheet],
      template: () => <ti-action-sheet alias={{ name: 'label' }} actions={options} />,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('点击遮罩关闭', async () => {
    const page = await newSpecPage({
      components: [TiActionSheet, TiPopup, TiTransition],
      template: () => <ti-action-sheet visible={true} title="标题" actions={actions} />,
    });
    const parent = page.root.shadowRoot.querySelector('ti-popup');
    const elm = parent.shadowRoot.querySelectorAll('ti-transition');
    elm[0].click();
    await page.waitForChanges();
  });

  it('点击菜单项', async () => {
    const page = await newSpecPage({
      components: [TiActionSheet, TiPopup, TiTransition],
      template: () => <ti-action-sheet visible={true} title="标题" actions={actions} />,
    });
    const parent = page.root.shadowRoot.querySelector('ti-popup');
    const btn = parent.querySelectorAll<HTMLDivElement>('.titian-action-sheet-button');
    btn[1].click();
    btn[2].click();
    await page.waitForChanges();
  });

  it('点击取消按钮', async () => {
    const page = await newSpecPage({
      components: [TiActionSheet, TiPopup, TiTransition],
      template: () => <ti-action-sheet visible={true} cancel-text="取消" title="标题" actions={actions} />,
    });
    const parent = page.root.shadowRoot.querySelector('ti-popup');
    const btn = parent.querySelector<HTMLButtonElement>('.titian-action-sheet-cancel');
    btn.click();
    await page.waitForChanges();
  });
});
