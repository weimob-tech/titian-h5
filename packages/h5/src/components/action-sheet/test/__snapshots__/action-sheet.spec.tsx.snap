// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`渲染 ti-action-sheet 基本渲染 1`] = `
<ti-action-sheet>
  <mock:shadow-root>
    <ti-popup class="titian-action-sheet" position="bottom">
      <div class="titian-action-sheet-title">
        标题
      </div>
      <div aria-hidden="true" class="titian-action-sheet-button">
        <div class="titian-action-sheet-text">
          <ti-badge attext="" use-slot="">
            <span>
              选项1
            </span>
          </ti-badge>
        </div>
        <div class="titian-action-sheet-description">
          描述1
        </div>
      </div>
      <div aria-hidden="true" class="titian-action-sheet-button">
        <div class="titian-action-sheet-text">
          <ti-icon extclass="titian-action-sheet-icon" name="home"></ti-icon>
          <ti-badge attext="" use-slot="">
            <span>
              选项1
            </span>
          </ti-badge>
        </div>
        <div class="titian-action-sheet-description">
          描述1
        </div>
      </div>
      <div aria-hidden="true" class="titian-action-sheet-button titian-action-sheet-disabled">
        <div class="titian-action-sheet-text">
          <ti-badge attext="" use-slot="">
            <span>
              选项1
            </span>
          </ti-badge>
        </div>
        <div class="titian-action-sheet-description">
          描述1
        </div>
      </div>
      <div aria-hidden="true" class="titian-action-sheet-button titian-action-sheet-loading">
        <div class="titian-action-sheet-text">
          <ti-loading extclass="titian-action-sheet-loading-icon" size="28"></ti-loading>
          <ti-badge attext="" use-slot="">
            <span>
              选项1
            </span>
          </ti-badge>
        </div>
        <div class="titian-action-sheet-description">
          描述1
        </div>
      </div>
    </ti-popup>
  </mock:shadow-root>
</ti-action-sheet>
`;

exports[`渲染 ti-action-sheet 字段别名 1`] = `
<ti-action-sheet>
  <mock:shadow-root>
    <ti-popup class="titian-action-sheet" position="bottom">
      <div aria-hidden="true" class="titian-action-sheet-button">
        <div class="titian-action-sheet-text">
          <ti-badge attext="" use-slot="">
            <span>
              选项1
            </span>
          </ti-badge>
        </div>
      </div>
    </ti-popup>
  </mock:shadow-root>
</ti-action-sheet>
`;
